packages/node/src/utils/kyve/kyve.ts

```diff
@@ -0,0 +1,465 @@
+        parseDecimal(currentBundle.id) > parseDecimal(b.id);
+      const isOutsiderBuffer =
+        height < parseDecimal(b.from_key) - bufferSize ||
+        height > parseDecimal(b.to_key) + bufferSize;
+
+      return isNotCurrentBundleAndLower && isOutsiderBuffer;
+    });
+  }
+
+  async clearFileCache(
+    cachedBundles: BundleDetails[],
+    height: number,
+    bufferSize: number,
+  ): Promise<void> {
+    const toRemoveBundles = this.getToRemoveBundles(
+      cachedBundles,
+      height,
+      bufferSize,
+    );
+
+    if (!toRemoveBundles.length) {
+      return;
+    }
+
+    for (const bundle of toRemoveBundles) {
+      const bundlePath = this.getBundleFilePath(bundle.id);
+      try {
+        await fs.promises.unlink(bundlePath);
+        remove(this.cachedBundleDetails, (b) => b.id === bundle.id);
+      } catch (e) {
+        if (e.code !== 'ENOENT') {
+          // if it does not exist, should be removed
+          throw e;
+        }
+      }
+    }
+  }
+
+  async getBlockByHeight(
+    height: number,
+  ): Promise<[BlockResponse, BlockResultsResponse]> {
+    const blocks = await this.fetchBundleCached(height);
+    const blockData = this.findBlockByHeight(height, blocks);
+
+    return [
+      this.decodeBlock(blockData.value.block),
+      this.injectLogs(this.decodeBlockResult(blockData.value.block_results)),
+    ];
+  }
+
+  private injectLogs(
+    kyveBlockResult: BlockResultsResponse,
+  ): BlockResultsResponse {
+    try {
+      kyveBlockResult.results.forEach((b) => {
+        // log is readonly hence needing to cast it
+        (b.log as any) = JSON.stringify(this.reconstructLogs(kyveBlockResult));
+      });
+    } catch (e) {
+      throw new Error(`Failed to inject kyveBlock, ${e}`);
+    }
+    return kyveBlockResult;
+  }
+
+  private reconstructLogs(
+    blockResultResponse: BlockResultsResponse,
+  ): logs.Log[] {
+    const logs: logs.Log[] = [];
+
+    for (const tx of blockResultResponse.results) {
+      let currentLog: any = {
+        msg_index: -1,
+        events: [],
+      };
+
+      let msgIndex = -1;
+      for (const event of tx.events) {
+        const isMessageEvent =
+          event.type === 'message' &&
+          event.attributes.some((e) => e.key === 'action');
+
+        if (isMessageEvent) {
+          if (msgIndex >= 0) {
+            logs.push(currentLog);
+          }
+          msgIndex += 1;
+          currentLog = { msg_index: msgIndex, events: [] };
+        }
+
+        if (msgIndex >= 0) {
+          currentLog.events.push(event);
+        }
+      }
+
+      if (currentLog.events.length > 0) {
+        logs.push(currentLog);
+      }
+    }
+    return logs;
+  }
+
+  async fetchBundleCached(height: number): Promise<KyveBundleData[]> {
+    let bundle = this.getBundleFromCache(height);
+
+    if (!bundle) {
+      const bundleId = await this.getBundleId(height);
+      bundle = await this.getBundleById(bundleId);
+      this.addToCachedBundle(bundle);
+    }
```

Comment:
I think the caching is redundant if this is here, there cant be anything that is awaited